core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/AbstractEvaluationStrategyFactory.java

```diff
@@ -55,4 +61,22 @@ public boolean isTrackResultSize() {
 	public void setTrackResultSize(boolean trackResultSize) {
 		this.trackResultSize = trackResultSize;
 	}
+
+	@Override
+	public void addOptimizer(QueryOptimizerFunctionalInterface queryOptimizerFunctionalInterface,
+			boolean beforeOtherQueryOptimizers) {
+		if (beforeOtherQueryOptimizers) {
+			queryOptimizersPre.add(queryOptimizerFunctionalInterface);
+		} else {
+			queryOptimizersPost.add(queryOptimizerFunctionalInterface);
+		}
+	}
+
+	public List<QueryOptimizerFunctionalInterface> getQueryOptimizersPre() {
+		return Collections.unmodifiableList(queryOptimizersPre);
+	}
+
+	public List<QueryOptimizerFunctionalInterface> getQueryOptimizersPost() {
+		return Collections.unmodifiableList(queryOptimizersPost);
+	}
```

Comment:
Why compile two separate collections (that then later on get concatenated into a single pipeline) instead of just immediately adjusting the pipeline itself?

Reply:
We don't have access to the pipeline yet. It doesn't get set until we instantiate an actual EvaluationStrategy. The pipeline in the EvaluationStrategyFactory can only be used if we don't want to use any of the standard optimizers.